### PR TITLE
JAMES-3775 RSpamD ADR

### PR DESCRIPTION
**Pull request by vttranlina**

### CODE REVIEW

File: src/adr/0055-rspamd-spam-filtering.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+## Status
+
+Accepted (lazy consensus).
+
+Implemented. 
+
+## Context
+
+In order to anti-spam, James provide 3 mechanism:
```
Comment (chibenwa) [inline src/adr/0055-rspamd-spam-filtering.md:13]:
```suggestion
In order to detect spam, James provide 3 mechanism: 
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+In order to anti-spam, James provide 3 mechanism: 
+
+- SMTP Hook: decide to reject or not before spooling on the SMTP level
+- Mailet: custom the mailet pipeline for changing the mail processing when detect spam mail.
+- Mailbox listener:  based on user action, eg: move in/out the message to Inbox/Spam mailbox, then decide report spam or not to separate Spam filtering system.
+
+For basic, we can base on username, DSN filter, domains, IP... for configuring black or grey list.
+If we want to analysis deeper in each message content more, we need a more complex system.
+The current, James supports the [SpamAssassin](https://spamassassin.apache.org/) for this problem.
```
Comment (chibenwa) [inline src/adr/0055-rspamd-spam-filtering.md:21]:
```suggestion
Currently, James is integrated with [SpamAssassin](https://spamassassin.apache.org/) to tackle this problem.
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+For more selection, James's repository provides the RSpamD extension, which provides the same way as SpamAssassin but another system - [RSpamD](https://github.com/rspamd/rspamd)
```
Comment (chibenwa) [inline src/adr/0055-rspamd-spam-filtering.md:23]: We should explain that we found hard to operate SpamAssassin, which we had performance issue with.

File: src/adr/0055-rspamd-spam-filtering.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+## Consequences
+
+- For higher performance, lower latency, the RSpamD should run in same network with James.
+- The query to RSpamD will get different score for same message. 
+
+
+## Alternatives
+
+- Rspamd can act as a [milter](https://en.wikipedia.org/wiki/Milter), we can use it to replace HTTP call, but more complicated
```
Comment (chibenwa) [inline src/adr/0055-rspamd-spam-filtering.md:62]:
```suggestion
- Rspamd can act as a [milter](https://en.wikipedia.org/wiki/Milter), we can use it to replace HTTP call, implementing a milter client in James is harder to implement.
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+- `RSpamDListener`: the listener will handler mailbox event, based on `MessageMoveEvent`, `MailboxEvents.Added` to detecting the mail is spam or ham, then report to RSpamD,
+en rich data to RSpamD, thus we will get more exact results in the next query.
+This listener will be configured in [mailetcontainer.xml](/server/apps/distributed-app/sample-configuration/listeners.xml).
+
+For connect to RSpamD, we use http protocol with reactor http client. 
+
+## Consequences
+
+- For higher performance, lower latency, the RSpamD should run in same network with James.
+- The query to RSpamD will get different score for same message.
```
Comment (chibenwa) [inline src/adr/0055-rspamd-spam-filtering.md:57]: Please mention a distributed mode for RSpamD is allowed by the use of Redis.

File: src/adr/0055-rspamd-spam-filtering.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+## Consequences
+
+- For higher performance, lower latency, the RSpamD should run in same network with James.
+- The query to RSpamD will get different score for same message. 
+- A distributed mode for RSpamD is allowed by the use of Redis.
+
+## Alternatives
+
+- Rspamd can act as a [milter](https://en.wikipedia.org/wiki/Milter), we can use it to replace HTTP call, implementing a milter client in James is harder to implement.
```
Comment (quantranhong1999) [inline src/adr/0055-rspamd-spam-filtering.md:63]:
```suggestion
- Rspamd can act as a [milter](https://en.wikipedia.org/wiki/Milter), we can use it to replace HTTP call. However, a milter client in James is harder to implement.
```

Sounds brighter to me.

File: src/adr/0055-rspamd-spam-filtering.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+## Status
+
+Accepted (lazy consensus).
+
+Implemented. 
+
+## Context
+
+In order to detect spam, James provide 3 mechanism:
```
Comment (Arsnael) [inline src/adr/0055-rspamd-spam-filtering.md:13]:
```suggestion
In order to detect spam, James provide 3 mechanisms: 
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+Implemented. 
+
+## Context
+
+In order to detect spam, James provide 3 mechanism: 
+
+- SMTP Hook: decide to reject or not before spooling on the SMTP level
+- Mailet: custom the mailet pipeline for changing the mail processing when detect spam mail.
+- Mailbox listener:  based on user action, eg: move in/out the message to Inbox/Spam mailbox, then decide report spam or not to separate Spam filtering system.
```
Comment (Arsnael) [inline src/adr/0055-rspamd-spam-filtering.md:17]:
```suggestion
- Mailbox listener:  based on user action, eg: move in/out the message to Inbox/Spam mailbox, then decide to report spam or not to separate Spam filtering system.
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+For basic, we can base on username, DSN filter, domains, IP... for configuring black or grey list.
+If we want to analysis deeper in each message content more, we need a more complex system.
```
Comment (Arsnael) [inline src/adr/0055-rspamd-spam-filtering.md:20]:
```suggestion
If we want to analyse deeper in each message content, we need a more complex system.
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+Currently, James is integrated with [SpamAssassin](https://spamassassin.apache.org/) to tackle this problem.
+We found it hard to operate SpamAssassin, and we had a performance issue.
+
+For more selection, James's repository provides the RSpamD extension, which provides the same way as SpamAssassin but another system - [RSpamD](https://github.com/rspamd/rspamd)
+
+A quick introduce about RSpamD:
```
Comment (Arsnael) [inline src/adr/0055-rspamd-spam-filtering.md:26]:
```suggestion
A quick introduction about RSpamD:
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+Rspamd is designed to process hundreds of messages per second simultaneously, and provides a number of useful features including a comprehensive Lua API that allows access to messages processing in various aspects as well as asynchronous network API to access external resources, such as DNS, HTTP or even generic TCP/UDP services.
+```
+
+## Decision 
+
+Set up a new maven project dedicated to rspamd extension. This allows to be embedded in a James server as a soft dependency
+using the external-jar loading mechanism. With this way, the extension could be dropped in one's James installation, and not a runtime dependency.
+
+Based on James' support customize mailets, listeners, web admin routes. RSpamD extension can be done via:
```
Comment (Arsnael) [inline src/adr/0055-rspamd-spam-filtering.md:41]:
```suggestion
Based on James' support for custom mailets, listeners, web admin routes, RSpamD extension can be done via:
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+- `RSpamDScanner` mailet: with each mail income, this mailet will query to RSpamD for get result spam or ham. Then append new headers to mail with status/flag spam.
```
Comment (Arsnael) [inline src/adr/0055-rspamd-spam-filtering.md:43]:
```suggestion
- `RSpamDScanner` mailet: with each mail income, this mailet will query to RSpamD for getting a spam or ham result, then append new headers to the mail with status/flag spam.
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+By setup with `IsMarkedAsSpam` matcher, the mail will be action rejected or not.
```
Comment (Arsnael) [inline src/adr/0055-rspamd-spam-filtering.md:44]:
```suggestion
By setting up with `IsMarkedAsSpam` matcher, the mail will be rejected or not.
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+This mailet will be configured in [mailetcontainer.xml](/server/apps/distributed-app/sample-configuration/mailetcontainer.xml).
+
+- Web admin route: to create feeding ham/spam messages task (batch mechanism). It helps spam classify learning.
+
+- `RSpamDListener`: the listener will handler mailbox event, based on `MessageMoveEvent`, `MailboxEvents.Added` to detecting the mail is spam or ham, then report to RSpamD,
```
Comment (Arsnael) [inline src/adr/0055-rspamd-spam-filtering.md:49]:
```suggestion
- `RSpamDListener`: the listener will handle mailbox events, based on `MessageMoveEvent`, `MailboxEvents.Added` to detect if the mail is spam or ham, then report to RSpamD,
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+en rich data to RSpamD, thus we will get more exact results in the next query.
```
Comment (Arsnael) [inline src/adr/0055-rspamd-spam-filtering.md:50]:
```suggestion
enrich data to RSpamD, thus we will get more exact results in the next query.
```

File: src/adr/0055-rspamd-spam-filtering.md [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+This listener will be configured in [mailetcontainer.xml](/server/apps/distributed-app/sample-configuration/listeners.xml).
+
+For connect to RSpamD, we use http protocol with reactor http client.
```
Comment (Arsnael) [inline src/adr/0055-rspamd-spam-filtering.md:53]:
```suggestion
To connect to RSpamD, we use http protocol with reactor http client. 
```

**Review by Arsnael — APPROVED** [review]

**Pull request merged**